redundant-bmc/src/services_impl.cpp

```diff
@@ -142,16 +161,119 @@ sdbusplus::async::task<> ServicesImpl::watchHostPropertiesChanged()
         auto it = properties.find("CurrentHostState");
         if (it != properties.end())
         {
-            auto hostState = std::get<HostState::HostState>(it->second);
-            poweredOn = util::getPoweredOnValue(hostState);
+            hostState = std::get<HostState::HostState>(it->second);
 
-            lg2::debug("Host state changed to {STATE}", "STATE", hostState);
+            lg2::debug("Host state changed to {STATE}", "STATE",
+                       hostState.value());
+
+            updateSystemState();
         }
     }
 
     co_return;
 }
 
+// NOLINTNEXTLINE
+sdbusplus::async::task<> ServicesImpl::readBootProgress()
+{
+    try
+    {
+        auto service = co_await util::getService(ctx, object_path::hostState,
+                                                 BootProgress::interface);
+        bootProgress = co_await BootProgress(ctx)
+                           .service(service)
+                           .path(object_path::hostState)
+                           .boot_progress();
+
+        lg2::debug("Initial BootProgress is {PROGRESS}", "PROGRESS",
+                   bootProgress.value());
+    }
+    catch (const sdbusplus::exception_t& e)
+    {
+        // Not on D-Bus yet
+    }
+
+    co_return;
+}
+
+// NOLINTNEXTLINE
+sdbusplus::async::task<> ServicesImpl::watchBootProgressPropertiesChanged()
+{
+    sdbusplus::async::match match(
+        ctx, rules::propertiesChanged(object_path::hostState,
+                                      BootProgress::interface));
+
+    while (!ctx.stop_requested())
+    {
+        auto [_, properties] = co_await match.next<std::string, HostPropMap>();
+
+        auto it = properties.find("BootProgress");
+        if (it != properties.end())
+        {
+            bootProgress = std::get<BootProgress::ProgressStages>(it->second);
+
+            lg2::debug("BootProgress changed to {PROGRESS}", "PROGRESS",
+                       bootProgress.value());
+
+            updateSystemState();
+        }
+    }
+
+    co_return;
+}
+
+void ServicesImpl::updateSystemState()
+{
+    using BProgress = BootProgress::ProgressStages;
+    using HState = HostState::HostState;
+
+    if (!hostState.has_value() || !bootProgress.has_value())
+    {
+        lg2::debug("Cannot calculate system state yet");
+        return;
+    }
+
+    SystemState newState = SystemState::other;
+
+    if (hostState == HState::Off)
+    {
+        newState = SystemState::off;
+    }
+    else if (hostState == HState::TransitioningToRunning)
+    {
+        newState = SystemState::booting;
+    }
+    else if (hostState == HState::Running)
+    {
+        if ((bootProgress.value() == BProgress::SystemInitComplete) ||
```

Comment:
The only other one which can sort of mean runtime is "SystemSetup". This is where we go when PHYP is booted but it's NVRAM has been cleared and you need to re-initialize it in the HMC. My vote would be we don't worry about it, but figured I'd put it out there in case you see something you want to handle there.

Reply:
As we discussed, I think I can leave out the check for that since it will always move one of those other 2 states afterwards.